### ISSUE TITLE
Als Squeak-Entwickler*in möchte ich sehen wie viel Speicher welche Methode meines Programms verbraucht, um zu verstehen, wo mein Programm Speicheroptimierungspotential hat. Ich möchte absolute und relativ Werte sehen.

### ISSUE THREAD

**Issue by linusha** [hpi-swa-teaching/InteractiveProfilingTool#18]

**Comment by tillprochaska** [comment]
Zusammenführen mit #15? Schätzung: 4h

**Comment by tillprochaska** [comment]
I think @twollnik and @Alexander-Dubrawski mentioned that it might be a little bit trickier to get the information on memory usage than runtime?

**Comment by twollnik** [comment]
> I think @twollnik and @Alexander-Dubrawski mentioned that it might be a little bit trickier to get the information on memory usage than runtime?

Yes, this ist true. Reading the printing methods of `MessageTally` may be a good starting point to find out how the memory stats should be interpreted.

Also we need to think of a way to switch between "memory-mode" and "time-mode". Maybe using a button?

Estimate: 6 h

**Comment by Alexander-Dubrawski** [comment]
All the memory relevant in the message Tally Objekt is saved in the Attribute `gcStats`. At first it is not obviously what the numbers in this array are representing. If we have a look in the method `MessageTally >> reportGCStatsOn:` we are able to get a first interpretation of the numbers.  

```smalltalk

	oldSpaceEnd			:= gcStats at: 1.
	youngSpaceEnd		:= gcStats at: 2.
	memoryEnd			:= gcStats at: 3.
```

*We can just imagine the memory ans an interval that is changing.*    
 
Trivially the youngSpaceEnd needs to be larger than the old Space because the space used is growing. So if we subtract the old space from the young space we get the value of the memory space this method needs.  
The hall memory used at the moment is the young space memory. 
The hole memory left on the machine is the memory end subtracted with youngSpaceEnd.

Now to the problem. In all object with exception of the tally tree root the `gcStats` attribute has the value nil. The size of the `gcStats` array in the root seams to be fixed as well with 73 entries. I think it is just possible to find the memory use for every method to change something in `MassageTally` class what I think isn't a trivial think. 

So I did some research. If we inspect the Method `MessageTally >> spyEvery:on:` we can see the following code:  

```smalltalk
	gcStats := Smalltalk getVMParameters.
```

What getVMParameters does is commented  as follow in the Method itself:  

*"Answer an Array containing the current values of the VM's internal parameter/metric registers.  Each value is stored in the array at the index corresponding to its VM register.  (See #vmParameterAt: and #vmParameterAt:put:.)" "Smalltalk getVMParameters"*   

In MessageTally this method is just called at the beginning of `spyEvery:on`. Are we calling this method for every MessageTally Object ? @twollnik @tillprochaska @linusha @nickpodratz   If not maybe we can just call this message every time we are creating a new MessageTally Object in the *MessageTally creation flow*. But before we try this I would ask Patrik or Eva, why the old profiler isn't saving the `gcStats` array for every method in the tree. Or if it is doing this, why it gets nil back.

So in my opinion I would ask Patrick first how often the profiler is calling `Smalltalk getVMParameters` and what are the consequences of doing so. After that I would be able to estimate  the time properly.

**Comment by linusha** [comment]
>Trivially the youngSpaceEnd needs to be larger than the old Space because the space used is growing. So if we subtract the old space from the young space we get the value of the memory space this method needs.
The hall memory used at the moment is the young space memory.
The hole memory left on the machine is the memory end subtracted with youngSpaceEnd.

Where did you get this information from? I did a bit of research in the Profiler-Repo from last time and found this:

>Subnodes show the change of used memory in different categories because of added/deleted objects during profiling. Memory space is divided into young space and old space. When an object is first created, it is put into young space. If an object has been in young space for quite a long time and during an incremental GC it is found that the object is still currently relevant to the executed process, it will be moved to old space. However most of the time, all references to an object will be gone fairly quickly so that that object "dies" in young space and never gets to reside in old space.

**Comment by Alexander-Dubrawski** [comment]
@linusha 
>>Where did you get this information from? I did a bit of research in the Profiler-Repo from last time and found this:

Please have a closer look at `MessageTally >> reportGCStatsOn:`. There is for example this section of the code : 

```smalltalk
str cr.
	str	nextPutAll: '**Memory**'; cr.
	str	nextPutAll:	'	old			';
		nextPutAll: (oldSpaceEnd asStringWithCommasSigned: true); nextPutAll: ' bytes'; cr.
	str	nextPutAll: '	young		';
		nextPutAll: ((youngSpaceEnd - oldSpaceEnd) asStringWithCommasSigned: true); nextPutAll: ' bytes'; cr.
	str	nextPutAll: '	used		';
		nextPutAll: (youngSpaceEnd asStringWithCommasSigned: true); nextPutAll: ' bytes'; cr.
	str	nextPutAll: '	free		';
		nextPutAll: ((memoryEnd - youngSpaceEnd) asStringWithCommasSigned: true); nextPutAll: ' bytes'; cr.
```

Based on this code I did my interpretation.  Moreover I didn't look at the old profiler repro. I looked at the profiler implementer in Squeak it self. But I don't see a conflict with my first interpretation and the information you found in the old repro. 

>>Could you please rephrase this or elaborate? I fear I am not getting your point.
please have a look at  MessageTally attribute `gcStats`. My Point is that we can't reconstruct the memory consumption for each note if the `gcStats` Attribute witch the only information about the memory consumption is nil.

>>But before we try this I would ask Patrik or Eva, why the old profiler isn't saving the gcStats array for every method in the tree. Or if it is doing this, why it gets nil back.

Like I said I I'm inspected the profiler in the Squeak image and not the old project. 
My intention was to get a better understanding of the profiling logic.

**Comment by Alexander-Dubrawski** [comment]
> > I think @twollnik and @Alexander-Dubrawski mentioned that it might be a little bit trickier to get the information on memory usage than runtime?
> 
> Yes, this ist true. Reading the printing methods of `MessageTally` may be a good starting point to find out how the memory stats should be interpreted.
> 
> Also we need to think of a way to switch between "memory-mode" and "time-mode". Maybe using a button?
> 
> Estimate: 6 h

I think a button would be a need feature, that can definitely improve the readability of the tree for the user. In some cases the runtime can be more important and in others the memory usage. I would suggest to create a different user story for this purpose. So Eve can decide if she wants this future. In all I find the idea good.

**Comment by twollnik** [comment]
@Alexander-Dubrawski Thank you for investing time in this. Please consider spellchecking your messages in the future. It is quite hard to understand you because of the spelling and grammar mistakes. I apologize for the inconvenience. (Note that I already made some edits to your earlier comments)

> I think a button would be a need feature

Do you mean a neat feature or a necessary feature?


@linusha 
>I thought the goal was to get the memory consumption as well as the runtime for each method call? In this case I am not sure for what we would need a button?

I fear that displaying both runtime and memory information may make it harder to understand the profiling results. I agree with Alex that Eva should be involved in this.

**Comment by linusha** [comment]
@Alexander-Dubrawski @twollnik 
I absolutely agree with both of you, that we need to discuss this with Eva if we were to implement this story! I would also agree, that it could become cluttered if we would display it both in the Tree Structure. I think the Profiler from SWT2017 had a nice way of doing this, with a small "details window" on the right side of the tree! 

@Alexander-Dubrawski definitely thanks for the reseach made by you, my goal was not to discredit this! Sorry if I made this impression!
I think that your interpretation of the different memory endings is correct __when just looking at the overall memory-consumption of the whole profiling run__ and this is exactly what the Profiler currently in the image does.
The Profiler from the 2017SWT does more, since they found a way to give the memory consumption for every Node in the tree. There the interpretations become more complicated, as I quoted above. But I agree with you that the solution to make memory-results available for every node needs some more research regarding the `getVMStats` method. I am positive that we can reverse-engineer how the past Profiler-Project did this, unfortunately, I currently do not have the time to provide more insight here.
